main.go

```diff
@@ -1,196 +1,625 @@
+		case "%s":`,
+			p)
+
+		srcPath := "src"
+		dstPath := "dst"
+
+		if sp := strings.Split(p, "."); len(sp) > 1 {
+			fd := md.GetMessageField(sp[0])
+			if fd.GetLabel() == descriptor.FieldDescriptorProto_LABEL_REPEATED {
+				panic(errors.New("Fieldmask for repeated field generated"))
+			}
+
+			goType := fieldTypeName(fd)
+
+			goName := generator.CamelCase(fd.GetName())
+			if v, ok := fd.OptionExtensions["gogoproto.customname"].(*string); ok {
+				goName = *v
+			}
+			if v, ok := fd.OptionExtensions["gogoproto.embed"].(*bool); ok && *v {
+				goName = goType
+			}
+
+			if fd.OneofIndex != nil {
+				oneOfType := fmt.Sprintf("%s_%s", md.GetName(), goName)
+				if md.GetMessage(goName) != nil {
+					oneOfType = fmt.Sprintf("%s_", oneOfType)
+				}
+				oneOfName := generator.CamelCase(md.GetOneofDecl()[fd.GetOneofIndex()].GetName())
+
+				dstPath = fmt.Sprintf("%s.%s", dstPath, oneOfName)
+
+				fmt.Fprintf(buf, `
+			if %s == nil {
+				%s = &%s{}
+			}`,
+					dstPath,
+					dstPath, oneOfType,
+				)
+
+				srcPath = fmt.Sprintf("%s.Get%s()", srcPath, goName)
+				dstPath = fmt.Sprintf("%s.(*%s).%s", dstPath, oneOfType, goName)
+			} else {
+				srcPath = fmt.Sprintf("%s.%s", srcPath, goName)
+				dstPath = fmt.Sprintf("%s.%s", dstPath, goName)
+			}
+
+			if v, ok := fd.OptionExtensions["gogoproto.nullable"].(*bool); ok && !*v {
+				fmt.Fprintf(buf, `
+			%s.SetFields(&%s, _pathsWithoutPrefix("%s", paths)...)`,
+					dstPath, srcPath, sp[0],
+				)
+
+			} else {
+				fmt.Fprintf(buf, `
+			if %s == nil {
+				%s = &%s{}
+			}
+			%s.SetFields(%s, _pathsWithoutPrefix("%s", paths)...)`,
+					dstPath,
+					dstPath, goType,
+					dstPath, srcPath, sp[0],
+				)
+			}
 			continue
 		}
-		sub.Name = proto.String(fmt.Sprintf("%s_%s", md.GetName(), sub.GetName()))
-		p.registerMessage(messages, sub)
+
+		fd := md.GetMessageField(p)
+
+		goName := generator.CamelCase(fd.GetName())
+		if v, ok := fd.OptionExtensions["gogoproto.customname"].(*string); ok {
+			goName = *v
+		}
+		if v, ok := fd.OptionExtensions["gogoproto.embed"].(*bool); ok && *v {
+			var goType string
+			switch fd.GetType() {
+			case descriptor.FieldDescriptorProto_TYPE_BOOL,
+				descriptor.FieldDescriptorProto_TYPE_DOUBLE,
+				descriptor.FieldDescriptorProto_TYPE_FLOAT,
+				descriptor.FieldDescriptorProto_TYPE_INT32,
+				descriptor.FieldDescriptorProto_TYPE_SINT32,
+				descriptor.FieldDescriptorProto_TYPE_SFIXED32,
+				descriptor.FieldDescriptorProto_TYPE_INT64,
+				descriptor.FieldDescriptorProto_TYPE_SINT64,
+				descriptor.FieldDescriptorProto_TYPE_SFIXED64,
+				descriptor.FieldDescriptorProto_TYPE_UINT32,
+				descriptor.FieldDescriptorProto_TYPE_FIXED32,
+				descriptor.FieldDescriptorProto_TYPE_UINT64,
+				descriptor.FieldDescriptorProto_TYPE_FIXED64,
+				descriptor.FieldDescriptorProto_TYPE_STRING,
+				descriptor.FieldDescriptorProto_TYPE_BYTES:
+				return nil, fmt.Errorf("Invalid type specifed for embedded field: %s", fd.GetType())
```

Comment:
1. our errors are lowercase
2. probably good to mention the containing message here instead of having something like `Invalid type specifed for embedded field: pointer`.